std/zig/parser.zig

```diff
@@ -125,6 +125,48 @@ pub const Parser = struct {
         FieldInitListItemOrEnd: ListState(&ast.NodeFieldInitializer),
         FieldInitListCommaOrEnd: ListState(&ast.NodeFieldInitializer),
         FieldListCommaOrEnd: &ast.NodeContainerDecl,
+
+        /// A state that can be appended before any other State. If an error occures,
+        /// the parser will first try looking for the closest optional state. If an
+        /// optional state is found, the parser will revert to the state it was in
+        /// when the optional was added. This will polute the arena allocator with
+        /// "leaked" nodes. TODO: Figure out if it's nessesary to handle leaked nodes.
+        Optional: Parser,
+
+        /// Optional can be reverted by adding the Required state to the stack.
+        Required,
```

Comment:
I don't know if I like this Optional and Required idea. It a footgun in so many ways, because we have no way of reverting the side effects other states have done. This means that `return 5 + ;` would actually store `5` as `return`s rhs. In this case that isn't a problem because there is an actual syntax error, but I'm sure there are corner cases I haven't thought of.